ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/PushQueryPublisher.java

```diff
@@ -63,32 +70,64 @@
     this.exec = requireNonNull(exec, "exec");
     this.query = requireNonNull(query, "query");
     this.localCommands = requireNonNull(localCommands, "localCommands");
+    this.pushRouting = null;
+    this.isScalablePush = false;
+    this.context = null;
+  }
+
+  PushQueryPublisher(
+      final KsqlEngine ksqlEngine,
+      final ServiceContext serviceContext,
+      final ListeningScheduledExecutorService exec,
+      final ConfiguredStatement<Query> query,
+      final PushRouting pushRouting,
+      final Context context
+  ) {
+    this.ksqlEngine = requireNonNull(ksqlEngine, "ksqlEngine");
+    this.serviceContext = requireNonNull(serviceContext, "serviceContext");
+    this.exec = requireNonNull(exec, "exec");
+    this.query = requireNonNull(query, "query");
+    this.localCommands = Optional.empty();
+    this.pushRouting = requireNonNull(pushRouting, "pushRouting");
+    this.isScalablePush = true;
+    this.context = requireNonNull(context, "context");
   }
 
   @Override
   public synchronized void subscribe(final Flow.Subscriber<Collection<StreamedRow>> subscriber) {
-    final TransientQueryMetadata queryMetadata = ksqlEngine
-        .executeQuery(serviceContext, query, true);
+    final PushQueryMetadata queryMetadata;
+    if (isScalablePush) {
+      final PushRoutingOptions routingOptions = new PushQueryConfigRoutingOptions(
+          ImmutableMap.of()
+      );
+
+      queryMetadata = ksqlEngine
+          .executeScalablePushQuery(serviceContext, query, pushRouting, routingOptions, context);
+    } else {
+      queryMetadata = ksqlEngine
+          .executeQuery(serviceContext, query, true);
+
+      localCommands.ifPresent(lc -> lc.write((TransientQueryMetadata) queryMetadata));
+    }
+
     final PushQuerySubscription subscription =
         new PushQuerySubscription(exec, subscriber, queryMetadata);
 
-    localCommands.ifPresent(lc -> lc.write(queryMetadata));
-
-    log.info("Running query {}", queryMetadata.getQueryApplicationId());
+    log.info("Running query {}", queryMetadata.getQueryId().toString());
```

Comment:
I started to follow the code here but it isn't super straightforward - what's the queryID of a SPQ? is it the same as the underlying persistent query? should we give it a meaningful name (especially if we want to have an admin api that allows users to terminate a single instance of an SPQ)?

Reply:
The query id is something like "SCALABLE_PUSH_QUERY_<TIMESTAMP>".  This will hopefully be used as you suggest.  We could potentially make the name more meaningful if we expose it in in this way.  I agree that adding the source makes sense. I can do that in a followup PR (maybe along with the admin API work) if that works.  What do you think?

Reply:
yeah, we can have this in a separate PR (definitely not urgent) we can file a ticket to track that as well (could be a good newbie task)